### ISSUE TITLE
Inaccurate Type Inference with `partial` and Type Arguments

### ISSUE THREAD

**Issue by ckwastra** [microsoft/pyright#7482]
When utilizing the `partial` function with a generic function as the first argument, Pyright successfully infers the return type. However, when the first argument is a type (e.g., `A`), Pyright only infers the return type as `A[Unknown]` instead of the expected `A[int]`.

**Minimal Reproducible Example**
Code sample in [Pyright Playground](https://pyright-play.net/?strict=true&code=GYJw9gtgBMCuB2BjALmMAbAzlAlhADmCMlPgIbE5noBQokUyAnvjvAOa4FEkDiApvH4gciADRQAKi34A1CjRqSoAXikz5IABQAiSToCUixOjKZsAQS0ChIxAG1JAXQMAuGlE9QAJv2BQAfQC2HGQgrUx%2BdGAJANcpAygAWgA%2BKAA5MCF4gDo8xV9-Mi04hOS0i0cnXPyaEH4AN35qAOZ8fi1ySmotCwk2ZC0DAyHEqABiKAAFJhF2AAsSNmBhTHjKgFV4AGt4MAB3eCcoLX4AD3aUfm91%2BwGXOsbm9FaZTopkKnQtMn74QeGo08kxmc0WuHgKxAayglXuJ0QRHqKHQTCMQA)

```python
from functools import partial
from typing import Generic, TypeVar

T = TypeVar("T")

class A(Generic[T]):
    def __init__(self, _: T) -> None: ...

def a(_: T) -> A[T]: ...

reveal_type(partial(A, int())())  # Pyright infers: A[Unknown] (expected: A[int])
reveal_type(partial(a, int())())  # Pyright infers: A[int] (correctly)

```

**Expected Outcome**
Pyright should infer the return type of `partial(A, int())()` as `A[int]` when `A` is a generic class, similar to how it correctly infers the return type when the first argument is a generic function.

**Comment by erictraut** [comment]
Pyright is working as designed here. (I'll note that its behavior is also consistent with mypy's behavior in this case.)

Object construction in Python involves a series of calls (first, to the metaclass `__call__` method, then to the class `__new__` method and finally to the class `__init__` method). There's significant logic involved in evaluating a call to a constructor. This is even more complex for generic classes because class-level and method-level TypeVars may need to be solved as part of this multi-call evaluation. 

When you pass a class as the first argument to `partial`, a type checker must convert that class into a single function call, effectively combining the `__new__` and `__init__` methods and solving the class-scoped TypeVars in the process. That's what's happening in your example above.

If you want to use a generic class constructor as the first argument to `partial`, you can explicitly specialize the class to get the desired effect.

```python
partial(A[int], int())()
```
